lisa/tools/wsl.py

```diff
@@ -0,0 +1,332 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import os
+import re
+import time
+from pathlib import PurePath
+from typing import TYPE_CHECKING, Any, Dict, Optional
+
+import charset_normalizer
+from assertpy.assertpy import assert_that
+
+from lisa.base_tools.wget import Wget
+from lisa.executable import Tool
+from lisa.util import LisaException, create_timer, find_groups_in_lines
+from lisa.util.process import ExecutableResult, Process
+
+from . import Echo, Find, Ls, PowerShell, Tar
+
+if TYPE_CHECKING:
+    from lisa.node import Node
+
+
+class Wsl(Tool):
+    """
+    Support commands on WSL host and WSL both.
+    """
+
+    FIXED_WSL_PATH = r"%ProgramFiles%\wsl\wsl.exe"
+    CONFIG_FILE_PATH = r"%USERPROFILE%\.wslconfig"
+
+    ENCODING = "utf-16-le"
+    LKG_RELEASE = "https://github.com/microsoft/WSL/releases/download/2.0.4/Microsoft.WSL_2.0.4.0_x64_ARM64.msixbundle"  # noqa: E501
+    INSTALL_TIMEOUT = 120
+
+    def __init__(self, node: "Node", guest: "Node") -> None:
+        assert guest, "guest node is required for Wsl tool."
+        super().__init__(node=node)
+        self._guest = guest
+        self._command_path = "wsl"
+
+    @property
+    def command(self) -> str:
+        return self._command_path
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def install(self) -> bool:
+        self._log.debug("wsl is not detected, or version is low. Installing...")
+        if self.node.is_remote:
+            self._install_on_remote()
+        else:
+            self._install_on_local()
+
+        return self._check_exists()
+
+    def install_distro(
+        self,
+        name: str,
+        reinstall: bool = False,
+        kernel: str = "",
+        enable_debug_console: bool = False,
+    ) -> None:
+        is_installed = False
+
+        if reinstall:
+            self._wsl_execute(
+                f"--unregister {name}",
+            )
+
+        # Ubuntu
+        # Ubuntu (Default)
+        distro_name_pattern = re.compile(rf"^\s*(?P<name>{name})\s*?.*?$")
```

Comment:
will it have different versions for the same distro? if yes, we can't just use name to know whether it has been installed or not

Reply:
The "Ubuntu" is the default name, and it indicates to 22.04 currently. There are versioned names like Ubuntu-18.04. The name is specified in runbook, and it can be set to reinstall, if the version needs to be confirmed.